Source/Core/VideoCommon/XFStructs.cpp

```diff
@@ -300,3 +300,511 @@ void PreprocessIndexedXF(u32 val, int refarray)
+    case 10:
+    case 11:
+    case 12:
+      // Yagcd says light pos or "inf ldir", while dolphin has a union for dpos and sdir with only
+      // dpos being used nowadays. As far as I can tell only the DX9 engine once at
+      // Source/Plugins/Plugin_VideoDX9/Src/TransformEngine.cpp used sdir directly...
+      return fmt::format("Light {0} {1} position or inf ldir {1} = {2}", light, "xyz"[offset - 10],
+                         reinterpret_cast<float&>(value));
+    case 13:
+    case 14:
+    case 15:
+      // Yagcd says light dir or "1/2 angle", dolphin has union for ddir or shalfangle.
+      // It would make sense if d stood for direction and s for specular, but it's ddir and
+      // shalfhangle that have the comment "specular lights only", both at the same offset,
+      // while dpos and sdir have none...
+      return fmt::format("Light {0} {1} direction or half hangle {1} = {2}", light,
+                         "xyz"[offset - 10], reinterpret_cast<float&>(value));
+    }
+  }
+  else
+  {
+    return fmt::format("Unknown memory {:04x} = {:08x}", address, value);
+  }
+}
+
+std::pair<std::string, std::string> GetXFTransferInfo(const u8* data)
+{
+  const u32 cmd = Common::swap32(data);
+  data += 4;
+  u32 baseAddress = cmd & 0xFFFF;
+  u32 transferSize = ((cmd >> 16) & 15) + 1;
+
+  if (baseAddress > 0x1058)
+  {
+    return std::make_pair("Invalid XF Transfer", "Base address past end of address space");
+  }
+  else if (transferSize == 1 && baseAddress >= 0x1000)
+  {
+    // Write directly to a single register
+    u32 value = Common::swap32(data);
+    return GetXFRegInfo(baseAddress, value);
+  }
+
+  // More complicated cases
+  std::string name, desc;
+  u32 endAddress = baseAddress + transferSize;  // exclusive
+
+  // do not allow writes past registers
+  if (endAddress > 0x1058)
+  {
+    name = "Invalid XF Transfer ";
+    desc = "Transfer ends past end of address space\n\n";
+    endAddress = 0x1058;
+  }
+
+  // write to XF mem
+  if (baseAddress < 0x1000)
+  {
+    u32 xfMemBase = baseAddress;
+    u32 xfMemTransferSize = transferSize;
+
+    if (endAddress > 0x1000)
+    {
+      xfMemTransferSize = 0x1000 - baseAddress;
+      baseAddress = 0x1000;
+    }
+
+    name += fmt::format("Write {} XF mem words at {:04x}", xfMemTransferSize, xfMemBase);
+
+    for (u32 i = 0; i < xfMemTransferSize; i++)
+    {
+      if (i > 0)
+        desc += '\n';
+      desc += GetXFMemDescription(xfMemBase + i, Common::swap32(data));
+      data += 4;
+    }
+
+    if (endAddress > 1000)
+    {
+      name += " ";
+    }
+  }
+
+  // write to XF regs
+  if (baseAddress >= 0x1000)
+  {
+    name += fmt::format("Write {} XF regs at {:04x}", endAddress - baseAddress, baseAddress);
+
+    for (u32 address = baseAddress; address < endAddress; address++)
+    {
+      u32 value = Common::swap32(data);
+
+      auto [regname, regdesc] = GetXFRegInfo(address, value);
+      desc += regname + "\n" + regdesc + "\n";
```

Comment:
```suggestion
      desc += fmt::format("{}\n{}\n", regname, regdesc);
```

So this doesn't churn string allocations.

Reply:
@lioncash - slightly off-topic but I've been curious about this for a while.  I wasn't able to find any documentation about fmt vs string concatenation in terms of memory allocations.

It makes sense to me for the line you mentioned.  Or even `fmt::format("{}\n", str);`.

But I'm curious, what about `fmt::format("{}{}", str1, str2);`?  Any idea on how fmt handles memory allocation?